### ISSUE TITLE
4.x Implementation Specs

### ISSUE THREAD

**Issue by Bird87ZA** [cakephp/cakephp#11365]
This is a (multiple allowed):

* [ ] bug
* [x] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 4.0.0
* Platform and Target: Any

Having worked with CakePHP a lot, I often find myself delving into the vendor code to see how something is implemented and if there could be any improvement. But CakePHP code is often hard to understand and sometimes hard to follow.

I'd like to suggest that the core team should perhaps look at following the principles (or good practices) as set out by 'Uncle Bob' in https://cleancoders.com/

Can we have a discussion on this and maybe think about this during implementation of new features and the process that needs to be followed when refactoring old code?

PS: Not sure what to name the discussion

**Comment by dereuromark** [comment]
Apart from https://github.com/cakephp/cakephp/issues/11260 I do not think that there is much missing here, to be honest. Sure, the DI topic is coming up once in a while as this is difficult to work with in a plugin context sometimes.

Also see [this](https://www.dereuromark.de/2016/01/29/developing-cakephp-3-plugins-its-fun/) on how you can already develop SOLID in CakePHP.

**Comment by markstory** [comment]
What kind of gaps have you found between the ideas in cleancoders and the code we have today?

**Comment by Bird87ZA** [comment]
Okay I've just opened ```cakephp/src/Controller/Controller.php``` and scrolled through but as I work with Cake I'll add to this discussion, but here are a few cases (even though it's not the reason for this post):

- A lot of underscored class variables and methods;
- Constructor is long with a lot of extractable code
- Line 260: ```$this->components()``` ? Is it setting or getting components? I know clicking through to it will make me understand what it does, but cleancoders state that if it's named correctly, that wouldn't be necessary.
- Redirect is extractable
- Render is extractable
- Referer is extractable (this method isn't that hard to understand, but for someone new to PHP, if this was to be a collection of methods that say exactly what each if statement does, it would be easier to follow.

I do understand that as a core member of cake, you might feel that this is unneeded and a useless addition, but not only will it make it easier for a newbie to adapt PHP as his/her framework of choice, it also enables the community to contribute more easily to the CakePHP if the code is easier to read.

**Comment by dereuromark** [comment]
> $this->components() ? Is it setting or getting components

We are actually in the process of building a clean API with clear setter and getter methods. This process takes time and needs deprecation usage to avoid BC breaks. But we are getting there.

> A lot of underscored class variables and methods;

Removing those is actually the total opposite of clean code. It prevents people especially in code review / pull request overview to quickly see the scope of this method and its visibility. That is why a few people actually would very much like to keep this very sane and useful guideline/standard.
Also note that PSR2 does not disallow it, it merely recommends not using them (which is a bad recommendation IMO).

> ... is extractable

Do you have any concrete ideas on how to do that? I think underlining your statement with code examples is the only way here, just pointing things out without showing how it might be cleaner does not help too much I am afraid.

As I stated above there is sure a bit too much code in general in the controller layer (I linked the related ticket for it), and we are sure happy for any helping hand in that process of getting things to the right place in the next minor and major versions.

**Comment by Bird87ZA** [comment]
> Removing those is actually the total opposite of clean code. It prevents people especially in code review / pull request overview to quickly see the scope of this method and its visibility. That is why a few people actually would very much like to keep this very sane and useful guideline/standard.
How is the scope of a method and its visibility not clearly visible by the access level? In my opinion, ```public```, ```private``` and ```protected``` should tell you what the scope and visibility of a method is regardless of what it is called. In the end, this boils down to preference, but I do however think it's an old custom (Badly implemented OO PHP4 days) that litters code for no real reason.
> Do you have any concrete ideas on how to do that? I think underlining your statement with code examples is the only way here, just pointing things out without showing how it might be cleaner does not help too much I am afraid.
As stated, as I come across them I'll post them. I was just looking at the milestones for CakePHP 4 when I thought of this, so didn't have any good examples on hand.
> We are actually in the process of building a clean API with clear setter and getter methods. This process takes time and needs deprecation usage to avoid BC breaks. But we are getting there.
This will probably solve most of the issues underlined by cleancoders.

As always, thanks for the response @dereuromark, always good to get a bit of a discussion going around improvements.

**Comment by inoas** [comment]
Underscored protected methods help with no-IDE situations especially if you then use double underscores for privates.

Anyway what I think is much more important is using const, final, private, signature type hints, return type hints, single-type returns, than a battle about underscore or not *shrug*.

**Comment by dereuromark** [comment]
@Bird87ZA Tell me please from [looking at this PR](https://github.com/cakephp/cakephp/pull/11363/files) what visibility the now used `getDependent()` has :)

**Comment by Bird87ZA** [comment]
@dereuromark 1:0 Bird.

From that PR I agree with you, using an underscore is easy to see private.

Doesn't mean I like it 😛

**Comment by inoas** [comment]
> Constructor is long with a lot of extractable code
> Redirect is extractable
> Render is extractable
> Referer is extractable (this method isn't that hard to understand, but for someone new to PHP, if this was to be a collection of methods that say exactly what each if statement does, it would be easier to follow).

I don't think that extracting things for the sake of extraction is a good things. Splitting code into a lot of method calls actually makes it harder to read NOT easier. Having logic blocks separated by a new line within one method is usually a lot easier to read, unless there is really some fancy complex logic going on - and even then one can revert to a closure with a meaningful name instead of a stand alone function.

Creating tons of interfaces, classes and methods will make readability, getting a "grasp" of what's going on, "from the outside", "as a newbie", *harder* not *easier*.

So I'd just say:
- Extract/refactor only when there is opportunity to go DRY.
- When function bodies grow, group code in chunks with new lines, use closures; if you must use methods within the same class/context. Spinning off chunks of code that only run once but are in some other class is really bad IMHO.

**Comment by ADmad** [comment]
> I don't think that extracting things for the sake of extraction is a good things. Splitting code into a lot of method calls actually makes it harder to read NOT easier.

Spot on. Having the code split across multiple small methods actually makes code difficult to follow.

Also methods calls are not "free". Having a lot of method calls per request eventually add a non trivial amount of additional processing time.

**Comment by mlingureanu** [comment]
There is a problem with __ methods in php:
```
Caution
PHP reserves all function names starting with __ as magical. It is recommended that you do not use function names with __ in PHP unless you want some documented magic functionality.
```

http://php.net/manual/en/language.oop5.magic.php

**Comment by dereuromark** [comment]
@mlingureanu Yes, that's why with protected you do not have that problem. For private you should never use `__`, but also stick to a single underscore (or avoid private alltogether).

**Comment by mlingureanu** [comment]
```
Warning
You cannot throw an exception from within a __toString() method. Doing so will result in a fatal error.
```
http://php.net/manual/en/language.oop5.magic.php#language.oop5.magic.tostring

**Comment by markstory** [comment]
@mlingureanu Is there additional context to your comment on `__toString()`? I don't understand how it is relevant to this issue.

**Comment by dereuromark** [comment]
All advice will be taken into consideration. Closing then this for now as the discussion came to a stop.
